dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/h12/grpc/GrpcCompositeCodec.java

```diff
@@ -16,123 +16,103 @@
 import org.apache.dubbo.remoting.http12.message.HttpMessageCodec;
 import org.apache.dubbo.remoting.http12.message.MediaType;
+import org.apache.dubbo.rpc.model.FrameworkModel;
+import org.apache.dubbo.rpc.model.MethodDescriptor;
+import org.apache.dubbo.rpc.model.PackableMethod;
+import org.apache.dubbo.rpc.model.PackableMethodFactory;
 
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.nio.charset.Charset;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
-import com.google.protobuf.Message;
-
-import static org.apache.dubbo.common.constants.CommonConstants.PROTOBUF_MESSAGE_CLASS_NAME;
+import static org.apache.dubbo.common.constants.CommonConstants.DEFAULT_KEY;
+import static org.apache.dubbo.common.constants.CommonConstants.DUBBO_PACKABLE_METHOD_FACTORY;
 
 public class GrpcCompositeCodec implements HttpMessageCodec {
 
-    private final ProtobufHttpMessageCodec protobufHttpMessageCodec;
+    private static final String PACKABLE_METHOD_CACHE = "PACKABLE_METHOD_CACHE";
 
-    private final WrapperHttpMessageCodec wrapperHttpMessageCodec;
+    private final URL url;
 
-    public GrpcCompositeCodec(
-            ProtobufHttpMessageCodec protobufHttpMessageCodec, WrapperHttpMessageCodec wrapperHttpMessageCodec) {
-        this.protobufHttpMessageCodec = protobufHttpMessageCodec;
-        this.wrapperHttpMessageCodec = wrapperHttpMessageCodec;
-    }
+    private final FrameworkModel frameworkModel;
+
+    private final String mediaType;
 
-    public void setEncodeTypes(Class<?>[] encodeTypes) {
-        this.wrapperHttpMessageCodec.setEncodeTypes(encodeTypes);
+    private PackableMethod packableMethod;
+
+    public GrpcCompositeCodec(URL url, FrameworkModel frameworkModel, String mediaType) {
+        this.url = url;
+        this.frameworkModel = frameworkModel;
+        this.mediaType = mediaType;
     }
 
-    public void setDecodeTypes(Class<?>[] decodeTypes) {
-        this.wrapperHttpMessageCodec.setDecodeTypes(decodeTypes);
+    public void loadPackableMethod(MethodDescriptor methodDescriptor) {
+        if (methodDescriptor instanceof PackableMethod) {
+            packableMethod = (PackableMethod) methodDescriptor;
+            return;
+        }
+        Map<MethodDescriptor, PackableMethod> cacheMap = (Map<MethodDescriptor, PackableMethod>) url.getServiceModel()
+                .getServiceMetadata()
+                .getAttributeMap()
+                .computeIfAbsent(PACKABLE_METHOD_CACHE, k -> new ConcurrentHashMap<>());
+        packableMethod = cacheMap.computeIfAbsent(methodDescriptor, md -> frameworkModel
+                .getExtensionLoader(PackableMethodFactory.class)
+                .getExtension(ConfigurationUtils.getGlobalConfiguration(url.getApplicationModel())
+                        .getString(DUBBO_PACKABLE_METHOD_FACTORY, DEFAULT_KEY))
+                .create(methodDescriptor, url, mediaType));
```

Comment:
**Security Issue**: The dynamic loading and caching mechanism for 'PackableMethod' instances does not appear to validate or sanitize the 'MethodDescriptor' or the URL used to retrieve configurations, which could potentially be exploited to inject malicious configurations or execute code remotely if external input is used to construct these objects. <br> **Fix**: Implement validation checks on 'MethodDescriptor' objects and any URLs or configurations used to create or retrieve 'PackableMethod' instances. Ensure that only trusted sources can influence these objects and that any input used in their creation is sanitized. <br> **Code Suggestion**: 
 ```
 +        if (methodDescriptor instanceof PackableMethod) {
 +            packableMethod = (PackableMethod) methodDescriptor;
 +            return;
 +        }
 +        // Validate methodDescriptor and URL before proceeding
 +        if (!isValidMethodDescriptor(methodDescriptor) || !isTrustedUrl(url)) {
 +            throw new IllegalArgumentException("Invalid MethodDescriptor or URL");
 +        }
 +        Map<MethodDescriptor, PackableMethod> cacheMap = (Map<MethodDescriptor, PackableMethod>) url.getServiceModel()
 +                .getServiceMetadata()
 +                .getAttributeMap()
 +                .computeIfAbsent(PACKABLE_METHOD_CACHE, k -> new ConcurrentHashMap<>());
 +        packableMethod = cacheMap.computeIfAbsent(methodDescriptor, md -> frameworkModel
 +                .getExtensionLoader(PackableMethodFactory.class)
 +                .getExtension(ConfigurationUtils.getGlobalConfiguration(url.getApplicationModel())
 +                        .getString(DUBBO_PACKABLE_METHOD_FACTORY, DEFAULT_KEY))
 +                .create(methodDescriptor, url, mediaType));
 ```